From 69083d593bb6b1fb345834970cacee2adaa2d4fa Mon Sep 17 00:00:00 2001
From: Stephan Gerhold <stephan.gerhold@linaro.org>
Date: Tue, 5 Aug 2025 17:44:32 +0200
Subject: [PATCH] remoteproc: qcom_q6v5: Send SMP2P stop signal in
 attached/detached state

If we stop the q6v5 remoteproc while it is in RPROC_ATTACHED or
RPROC_DETACHED state, we still want to send the stop signal to shut it down
cleanly.

The main goal of the check in qcom_q6v5_request_stop() is to avoid sending
duplicate shutdown/stop signals during crash or shutdown via sysmon, so
check for RPROC_CRASHED to handle all of RPROC_RUNNING, RPROC_ATTACHED and
RPROC_DETACHED. It does not make sense to call the function while in
RPROC_OFFLINE state.

On Linux 7.2.9, qcom_q6v5_request_stop() uses the running flag instead
of the remoteproc state, preserving stop requests for logical crashes
while skipping watchdog/fatal shutdowns. Mark the boot firmware as running
when SMP2P detects a detached remoteproc so that its initial stop sends the
SMP2P signal too. Attached remoteprocs already set this flag in q6v5_init().

Signed-off-by: Stephan Gerhold <stephan.gerhold@linaro.org>
---
 drivers/remoteproc/qcom_q6v5.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/remoteproc/qcom_q6v5.c b/drivers/remoteproc/qcom_q6v5.c
index 3c173a4fc..c177780bf 100644
--- a/drivers/remoteproc/qcom_q6v5.c
+++ b/drivers/remoteproc/qcom_q6v5.c
@@ -289,8 +289,10 @@ void qcom_q6v5_read_smp2p_state(struct qcom_q6v5 *q6v5)
 	irq_get_irqchip_state(q6v5->fatal_irq, IRQCHIP_STATE_LINE_LEVEL, &fatal);
 	irq_get_irqchip_state(q6v5->stop_irq, IRQCHIP_STATE_LINE_LEVEL, &stop);
 
-	if (ready && handover && !stop && !fatal)
+	if (ready && handover && !stop && !fatal) {
 		q6v5->rproc->state = RPROC_DETACHED;
+		q6v5->running = true;
+	}
 }
 EXPORT_SYMBOL_GPL(qcom_q6v5_read_smp2p_state);
 
-- 
2.55.0

